File: api/src/main/java/com/ford/labs/retroquest/column/ColumnService.java
/*
 * Copyright (c) 2022 Ford Motor Company
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ford.labs.retroquest.column;

import com.ford.labs.retroquest.exception.ColumnNotFoundException;
import com.ford.labs.retroquest.websocket.WebsocketService;
import com.ford.labs.retroquest.websocket.events.WebsocketColumnEvent;
import com.ford.labs.retroquest.websocket.events.WebsocketEventType;
import io.micrometer.core.instrument.MeterRegistry;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ColumnService {

    private final ColumnRepository columnRepository;
    private final MeterRegistry meterRegistry;
    private final WebsocketService websocketService;

    public ColumnService(ColumnRepository columnRepository, MeterRegistry meterRegistry, WebsocketService websocketService) {
        this.columnRepository = columnRepository;
        this.meterRegistry = meterRegistry;
        this.websocketService = websocketService;
    }

    public List<Column> getColumns(String teamId) {
        return columnRepository.findAllByTeamId(teamId).stream().sorted().toList();
    }

    public Column editTitle(Long columnId, String title, String teamId) {
        var existingColumn = fetchColumn(teamId, columnId);
        existingColumn.setTitle(title);

        Column newColumn = columnRepository.save(existingColumn);

        websocketService.publishEvent(new WebsocketColumnEvent(teamId, WebsocketEventType.UPDATE, newColumn));

        meterRegistry.counter("retroquest.columns.changed.count").increment();

        return newColumn;
    }

    public Column fetchColumn(String teamId, Long columnId) {
        return columnRepository.findByTeamIdAndId(teamId, columnId).orElseThrow(ColumnNotFoundException::new);
    }
}

File: ui/src/Services/Api/ColumnService.ts
/*
 * Copyright (c) 2022 Ford Motor Company
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import axios, { AxiosResponse } from 'axios';

import { Column } from '../../Types/Column';

import getAuthConfig from './getAuthConfig';

const ColumnService = {
	getColumns(teamId: string): Promise<Column[]> {
		const url = `/api/team/${teamId}/columns`;
		return axios.get(url, getAuthConfig()).then((response) => response.data);
	},

	updateTitle(
		teamId: string,
		columnId: number,
		title: string
	): Promise<AxiosResponse<void>> {
		const url = `/api/team/${teamId}/column/${columnId}/title`;
		return axios.put(url, { title }, getAuthConfig());
	},
};

export default ColumnService;
